app/src/main/java/de/hhn/aib/labsw/blackmirror/AbstractActivity.kt

```diff
@@ -0,0 +1,81 @@
+package de.hhn.aib.labsw.blackmirror
+
+import androidx.appcompat.app.AppCompatActivity
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.JsonNode
+import de.hhn.aib.labsw.blackmirror.controller.API.Websockets.ApiListener
+import de.hhn.aib.labsw.blackmirror.controller.API.Websockets.MirrorApi
+import de.hhn.aib.labsw.blackmirror.controller.API.Websockets.MirrorApiWebsockets
+import okhttp3.OkHttpClient
+import okhttp3.Request
+
+
+abstract class AbstractActivity: AppCompatActivity(),ApiListener {
+    /**
+     * call this method to send data to the mirror
+     * @param topic topic of the message, not null
+     * @param payload payload of the message, not null
+     */
+    protected open fun publish(topic:String, payload: JsonNode){
+        api.publish(topic,payload)
+    }
+
+    /**
+     * call this method to send data to the mirror
+     * @param topic topic of the message, not null
+     * @param payload payload of the message, not null
+     */
+    protected open fun publish(topic:String, payload:Any){
+        api.publish(topic,payload)
+    }
+
+    /**
+     * call this method to subscribe to a topic
+     * @param topic topic to subscribe, not null
+     * @param listener listener to receive updates, not null
+     */
+    protected open fun subscribe(topic:String, listener: ApiListener){
+        api.subscribe(topic,listener)
+    }
+
+    /**
+     * call this method to revoke a subscription.
+     * only the subscription to the given topic will be removed. All other remain active
+     * @param topic topic of the subscription, not null
+     * @param listener listener to be removed, not null
+     */
+    protected open fun unsubscribe(topic:String, listener: ApiListener){
+        api.unsubscribe(topic,listener)
+    }
+
+    /**
+     * use this method to convert JsonNode->class instance
+     * @param node node to be converted, not null
+     * @param tClass class of which the instance should be created, not null
+     */
+    @Throws(JsonProcessingException::class)
+    protected fun <T> nodeToObject(node: JsonNode, tClass: Class<T>): T {
+        return api.getJSONMapper().treeToValue(node, tClass)
```

Comment:
would be replaced with property access, see comment on MirrorApiWebsockets::getJsonMapper()